doc/clicc.1
.\" -------------------------------------------------------------------------
.\" Projekt : CLICC - a Common Lisp to C Compiler
.\"           -----------------------------------
.\" Function : The CLICC Manual Page
.\"
.\" $Source: /home3/apply/public/clicc-0.6.4/doc/RCS/clicc.1,v $
.\" $Author: hk $
.\" $Revision: 1.3 $
.\" $Date: 1993/12/09 15:04:30 $
.\" -------------------------------------------------------------------------
.\"
.TH CLiCC 1 "30 Jul 1993"
.SH NAME
clicc \- Common Lisp to C compiler
.SH SYNOPSIS
within Common Lisp:

(clicc <filename>
.PD 0
.IP
.I :verbose
<T_or_NIL>
.IP
.I :print
<T_or_NIL>
.IP
.I :memsizes
<memsizes>
.IP
.I :no-codegen
<T_or_NIL>
.IP
.I :ti-level        
<0..3>
.IP
.I :module-compiler 
<T_or_NIL>
.IP
.I :out             
<filename>
.IP
.I :max-lines       
<count>
.IP
.I :split           
<T_or_NIL>
.IP
.I :flat-ifs        
<T_or_NIL>)

.SH DESCRIPTION
.I CLiCC
is a Common Lisp to C Compiler.  It generates C-executables 
from Common Lisp application programs. 
.I CLiCC 
is not a Common Lisp 
system.  Hence it does  not  include any  program  development or 
debugging support.  
.I CLiCC 
is intended to  be used as an  addon to 
existing Common Lisp systems for generating portable applications.
 
.I CLiCC 
supports a subset of Common Lisp + CLOS, which we call CL_0
(CommonLisp_0).  CL_0  is a strict and very large  subset of full
Common Lisp + CLOS, without  (EVAL ...) and friends.  At present,
CL_0 is based on CLtL1,  but  we  are  working towards  CLtL2 and
ANSI-CL.
 
The target language is a subset of C. 
.I CLiCC 
can be configured to generate either ANSI-C or K&R-C.  The generated C
code is compilable using a conventional C compiler on the target
machine, and must be linked with the
.I CLiCC 
runtime library in order to generate executables.

.SH OPTIONS
.TP
.IR ":verbose   " <T_or_NIL>
Output initial logo  (
.I T
)  or supress it  (
.I NIL
).

.TP
.IR ":print    " <T_or_NIL>
Output lots of progress messages  (
.I T
)  or keep
quiet  (
.I NIL
).

.TP
.IR ":memsizes    " <memsizes>
Determines the minimal and maximal sizes for heap and stack that are
initially allocated by the generated program. 
<memsizes> has the form  
.IP
.I ((stack-min . stack-max) 
.IP
.I (formheap-min . formheap-max) 
.IP
.I (fixnumheap-min . fixnumheap-max) 
.IP
.IR "(floatheap-min . floatheap-man))".
.IP 
All elements of this data structure are fixnums. At startup time the
compiled program tries to allocate the maximal amount of memory for
each of its memory areas. If not that much memory is available it
succesively halfs its requirements until they are satisfied.  If the
available memory falls below the minimal amount, the program gives up.

.TP
.IR ":no-codegen     " <T_or_NIL>
Do not generate code, only run all prior passes  (
.I T
)  or generate object (= C) code  (
.I NIL
).

.TP
.IR ":ti-level      "  <0..3>
Controls the intensity of type inference, which
.I CLiCC 
should perform.
.I ti-level
ranges from 0  (no type inference)  to 3  (exhaustive type inference).
.I ti-level
3 can take quite a long time on medium or large size programs.
 
.TP
.IR ":module-compiler   " <T_or_NIL>
.I CLiCC
acts as a module compiler  (
.I T
)  or as an application compiler which compiles whole programs  (
.I NIL
). The module compiler allows seperate compilation of Lisp modules. 

.TP
.IR ":out " <filename>
<filename> is the name of the file 
.I CLiCC
will generate from its input.
The default is the same name as the input file but with extension ".c"

.TP
.IR ":max-lines   "  <count>
if 
.I :split
-mode is off, 
.I :max-lines
limits the maximal number of lines in generated C-files to <count>. If
<count> has the value 
.I NIL
no spliting of files takes place.

.TP
.IR ":split       " <T_or_NIL>
generate a C-file for each function  (
.I T
)  or put all generated code in one file  (
.I NIL
) according to the 
.I :max-lines
option.

.TP 
.IR ":flat-ifs    " <T_or_NIL>)
controls the generated C-code for conditionals.
Generate either nested C-ifs  ( 
.I T
)  or flat C-ifs with gotos. Some C-Compiler cannot compile highly
nested control structures.


.SH REFERENCE

The language implemented conforms to

.RS 1
      Heinz Knutzen: Common Lisp_0 Sprachdefinition.
      APPLY technical report APPLY/CAI/II.2/1.	
      CAU Kiel, April 1993
.RE

it contains the negative diffs relative to

.RS 1
      Guy L. Steele Jr.: Common Lisp - The Language.
      Digital Press. 1st edition 1984, 465 pages.
      ("CLtL1" for short)
.RE

.SH ENVIRONMENT
.TP
.BR CLICCROOT
is used for determining the base directory of 
.I CLiCC 
specific files.

.SH FILES
All file locations are relative to the 
.I CLiCC
base directory:
.TP
.I README INSTALL COPYRIGHT
if you haven't read them do so now.
.TP
.I doc/*
numerous documentation files
.TP
.I src/compiler/config.lisp
.I src/compiler/clcdef.lisp
various configuration parameters
.TP
.I src/compiler/*.lisp
.I CLiCC
compiler source code
.TP
.I src/runtime/c/*.c
.I CLiCC
kernel run time system source code
.TP
.I src/runtime/lisp/*.lisp
.I CLiCC 
run time system source code 
.TP
.I src/runtime/lisp2c/*.c
.I CLiCC 
run time system compiled from LISP to C

.SH SEE ALSO
.IR cmulisp (1),
.IR clisp (1),
.IR cc(1).

.SH PROJECTS
More work on type inference and related optimizations; improving the
object oriented part; including a mean and lean foreign
function interface for integration with the C world.

.SH AUTHORS
.HP 0
Wolfgang Goerigk, wg@informatik.uni-kiel.de
.HP 0
Ulrich Hoffmann, uho@informatik.uni-kiel.de
.HP 0
Heinz Knutzen, hk@informatik.uni-kiel.de

.SH MAIL ADDRESS

.PD 0
.HP 0
Institut fuer Informatik und Praktische Mathematik
.HP 0
Preusserstr. 1 - 9
.HP 0
24105 Kiel, Germany

This work was supported by the German Federal Ministry for Research
and Technology (BMFT) within the joint project APPLY .




src/compiler/clicc.lisp
;;;-----------------------------------------------------------------------------
;;; Copyright (C) 1993 Christian-Albrechts-Universitaet zu Kiel, Germany
;;;-----------------------------------------------------------------------------
;;; Projekt  : APPLY - A Practicable And Portable Lisp Implementation
;;;            ------------------------------------------------------
;;; Funktion : Hauptprogramm des Compilers
;;;
;;; $Revision: 1.16 $
;;; $Log: clicc.lisp,v $
;;; Revision 1.16  1994/05/22  15:07:49  sma
;;; Neuer Kommandoswitch -R zum Einstellen der Datenrepräsentation.
;;;
;;; Revision 1.15  1994/05/19  07:57:02  pm
;;; Fehler behoben
;;;
;;; Revision 1.14  1994/05/17  08:21:28  pm
;;; Abgleich der FFI-Exports mit clcload.lisp
;;;
;;; Revision 1.13  1994/04/22  14:09:54  pm
;;; Foreign Function Interface voellig ueberarbeitet.
;;; - Ueberfluessiges exportiertes Symbol entfernt
;;;
;;; Revision 1.12  1994/04/18  12:05:56  pm
;;; Foreign Function Interface voellig ueberarbeitet.
;;; - Liste der exportierten Symbole ergaenzt
;;;
;;; Revision 1.11  1993/12/21  09:29:59  hk
;;; Schreibfehler in Usage behoben.
;;;
;;; Revision 1.10  1993/12/06  16:42:15  hk
;;; Flag geändert und ergänzt: -v verbose, -V version, -O opt cycles.
;;; Aufgeräumt. Gemischte Flags -vVo <name> gehen nun. Initialisierung
;;; globaler Variablen nach clcdef.
;;;
;;; Revision 1.9  1993/10/26  11:32:34  pm
;;; Abgleich der FFI-Exports mit clcload.lisp
;;;
;;; Revision 1.8  1993/06/17  08:00:09  hk
;;; Copright Notiz eingefuegt
;;;
;;; Revision 1.7  1993/06/09  12:45:09  hk
;;; An den Modulcompiler angepasst, Packages mit clcload konsistent
;;; gemacht.
;;;
;;; Revision 1.6  1993/04/06  09:59:11  hk
;;; Nickname L.
;;;
;;; Revision 1.5  1993/02/19  13:58:46  hk
;;; :use () beim LISP Package.
;;;
;;; Revision 1.4  1993/02/16  15:27:00  hk
;;; Revision Keyword eingefuegt.
;;;
;;; Revision 1.3  1993/01/11  16:48:37  hk
;;; argv -> c-argv
;;;
;;; Revision 1.2  1992/08/07  11:40:36  hk
;;; Dateikopf verschoenert.
;;;
;;; Revision 1.1  1992/03/24  16:54:56  hk
;;; Initial revision
;;; 
;;; Changes  : 29.10.91 Bedienung durch Kommandozeilenparameter 
;;;-----------------------------------------------------------------------------

(in-package "LISP" :nicknames '("CLICC-LISP" "L") :use ())
(in-package "USER")                     ; Damit es dieses Package wirklich gibt
(lisp:in-package "RT" :nicknames '("RUNTIME") :use ())
(lisp:in-package "FFI" :use ())
(lisp:export
 '(c-array c-char c-char-p c-double c-double-p c-enum c-float c-float-p c-fun
   c-handle c-int c-int-p c-long c-long-double c-long-double-p c-long-p c-ptr
   c-short c-short-p c-string c-string-p c-struct c-union c-unsigned-char
   c-unsigned-char-p c-unsigned-int c-unsigned-int-p c-unsigned-long
   c-unsigned-long-p c-unsigned-short c-unsigned-short-p c-vararg c-void
   copy-c-cstring def-c-type def-call-in def-call-out foreign-package-name
   free lisp-character lisp-float lisp-integer load-foreign make-c-string
   make-lisp-string)
 "FFI")

(lisp:in-package "CLICC" :use '("LISP"))

;;------------------------------------------------------------------------------
(require "clcmain")

(defconstant *COMMAND-OPTIONS* '((#\V . *SHOW-VERSION*)
                                 (#\v . *CLICC-PRINT*)
                                 (#\m . *MODULE-COMPILER*)
                                 (#\l . *LISP-MODULE*)
                                 (#\i . *INLINE-MODULE*)
                                 (#\s . *SPLIT-FILES*)
                                 (#\f . *FLAT-IFS*)
                                 (#\t *TI-LEVEL*)
                                 (#\R *OBREP*)
                                 (#\O *ITERATIONS*)
                                 (#\c *C-max-line-count*)
                                 (#\o *OUT-FILENAME*)))



(defun clicc-main ()
  (labels
      ((usage ()
         (format
          t
"Usage: clicc [-lmsvV] [-c count] [-t 0-3] [-O level] [-R 1-3] [-o name] name~%~
                  ~A-c: Maximum line count of C file~%~
                  ~A-f: Flat IFs~%~
                  ~A-i: Inline Module~%~
                  ~A-l: Lisp Module~%~
                  ~A-m: Module-Compiler~%~
                  ~A-O: Number of optimization cycles~%~
                  ~A-s: Split Files~%~
                  ~A-t: Type Inference Level 0, 1, 2 or 3~%~
                  ~A-R: Data Representation schema 1, 2 or 3~%~
                  ~A-v: Verbose~%~
                  ~A-V: Show Version~%~
                  ~A-o: Output file~%~
                  ~Aname: The Lisp File~%"
#\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab #\Tab)
         (return-from clicc-main)))
       
    (setq *MODULE-COMPILER* nil
          *LISP-MODULE* nil
          *INLINE-MODULE* nil
          *OUT-FILENAME* nil
          *FILENAME* nil)

    (let ((options (cdr (rt::c-argv))))
      (when (null options)
        (usage))
      (do ((opt (car options) (car options)))
          ((endp options)
           (unless *FILENAME* (usage))
           (when (stringp *ITERATIONS*)
             (setq *ITERATIONS* (read-from-string *ITERATIONS*))
             (unless (integerp *ITERATIONS*) (usage))
             (when (zerop *ITERATIONS*) (setq *optimize* nil)))
           (when (stringp *TI-LEVEL*)
             (let ((l (digit-char-p (character *TI-LEVEL*))))
               (if l (setq *TI-LEVEL* l) (usage))))
           (when (stringp *OBREP*)
             (let ((o (digit-char-p (character *OBREP*))))
               (if (and o (<= 1 o 3)) (setq *OBREP* o) (usage))))
           (when (stringp *C-max-line-count*)
             (setq *C-max-line-count* (read-from-string *C-max-line-count*))
             (unless (integerp *C-max-line-count*) (usage)))
           (do-clicc))
        (pop options)
        (cond
          ((eql #\- (char  opt 0))
           (dotimes (i (1- (length opt)))
             (let ((var (cdr (assoc (char opt (1+ i)) *COMMAND-OPTIONS*)))) 
               (if var
                   (if (atom var)
                       (set var t)
                       (if (consp options)
                           (setf (symbol-value (car var)) (pop options))
                           (usage)))
                   (usage)))))
          ((not *FILENAME*) (setq *FILENAME* opt))
          (T (usage)))))))

(clicc-main)
